lib/puppet/pops/validation/checker4_0.rb

```diff
@@ -530,6 +533,88 @@ def internal_check_future_reserved_word(o, name)
     end
   end
 
+  def internal_check_file_scope(o, name, file)
+    return if file.nil?
+
+    file_scope = @file_scope_hash[file]
+    if file_scope.nil?
+      return if @file_scope_hash.key?(file)
+      file_scope = @file_scope_hash[file] = generate_file_scope file
+      return if file_scope.nil?
+    end
+
+    if check_scope(name, file_scope)
+      acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    end
+    #Return to direct call after benchmarking check method
+    #if name.start_with? file_scope
+    #if name !~ file_scope
+    #  acceptor.accept(Issues::ILLEGAL_DEFINITION_LOCATION, o, {:name => name, :file => file})
+    #end
+  end
+
+  def create_scope_match(names)
+    qualified_name = names.join("::")
+
+    #qualified_name
+    Regexp.new "^#{qualified_name}", Regexp::IGNORECASE
+  end
+
+  def check_scope(name, file_scope)
+    #!name.start_with?(file_scope)
+    name !~ file_scope
+  end
+
+  def generate_file_scope(file)
```

Comment:
If changing the term "scope" to "namespace", this method could change to `namespace_for_file(file)`